### PR TITLE
JAVA-2265: Configure NullSavingStrategy at the DAO level

### PR DESCRIPTION
**Pull request by tomekl007**
…or Update annotation

### CODE REVIEW

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: This is code responsible for finding out if the `nullSavingStrategy` value was explicitly set on the `@Update` annotation using `AnnotationMirror`. It looks a bit complex, @olim7t do you know if there is a way to create `ExecutableElement` using java-poet for the purpose of this test?

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (tomekl007) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: I think If we choose to add new annotation `DefaultNullSavingStrategy` this annotation should have `nullSavingStrategy` without default. If someone is adding that at the `@Dao` level it means that he must specify `nullSavingStrategy` that will apply to inherited strategy for methods: `@Update`, `@Insert`, `@SetEntity`, `@Query`.

**Review by tomekl007 — COMMENTED** [review]
Implemented only for `@Update` and waiting for feedback.
`@SetEntity`, `@Insert` and `Query` level handing todo (it is fairly simple: code from line https://github.com/datastax/java-driver/pull/1246/files#diff-25ba2cdaf8c214d0d4a6f05f1c7922acR129 to https://github.com/datastax/java-driver/pull/1246/files#diff-25ba2cdaf8c214d0d4a6f05f1c7922acR135)

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: I agree with no default.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: JavaPoet only generates strings. The only way to create an `ExecutableElement` for a test is either through mocking or with a full compiler execution I think.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: You can compare the type with a `Class` directly:
```java
context.getClassUtils().isSame(getAnnotationType(), Update.class)
```

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: From `getElementValues()` javadocs:
> Only those elements with values explicitly present in the annotation are included, not those that are implicitly assuming their default values.

So we just need to iterate `getElementValues().keySet()` and find out if there is an element such that `key.getSimpleName().contentEquals("nullSavingStrategy")`.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: I think the Stream API really makes things worse here (and to be honest that's true most of the time -- that's why it's discouraged in the contribution guidelines).

There's an imperative version for a similar case [here](https://github.com/datastax/java-driver/blob/java2078/mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoQueryProviderMethodGenerator.java#L84). It's not pretty either, but I find it way more readable.

Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: I've changed (https://github.com/datastax/java-driver/pull/1246/commits/75a978baf7118914f2907e47e89a46ead8e7561f) it to imperative way but I am not sure if this is more readable that stream versions after simplification:
```
  return methodElement
        .getAnnotationMirrors()
        .stream()
        .filter(v -> classUtils.isSame(v.getAnnotationType(), javaClass))
        .flatMap(v -> v.getElementValues().keySet().stream())
        .anyMatch(v -> v.getSimpleName().contentEquals("nullSavingStrategy"));
```

**Comment by tolbertam** [comment]
Sorry for delay in review, this will be my first priority tomorrow :+1:

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: You could move this to a static method on `DaoBase`, so that you only have to generate the method call.

On the other hand the code is still relatively simple so it's not a big deal to generate it. Your call.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (olim7t) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java]: I think we don't need a dedicated parameter name here, since it's the only one and it's pretty obvious from the name of the annotation. So I would call this method `value()` so that it can be written directly like this:

```java
@DefaultNullSavingStrategy(SET_TO_NULL)
```

File: manual/mapper/daos/defaultnullsavingstrategy/README.md
Comment (olim7t) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: Can you make `@DefaultNullSavingStrategy` a link to the Javadocs?

File: manual/mapper/daos/null_saving/README.md
Comment (olim7t) [inline manual/mapper/daos/null_saving/README.md]: I don't think a separate page is necessary here, I would inline it in this section.
It fits in nicely since it relates to the main topic.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:77]: 👍 Good idea to extract the boilerplate into a separate component

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java]: Maybe calling this `daoNullSavingStrategy` would make the code below a bit easier to read?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: We might want to make this class (and its two package-private methods) public.
We're not committing to the API because it's in an internal package, and it could come in handy if we extend the generators or create new ones from the DSE mapper.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:98]: This is a bit hard to follow, but I think it's more because of the inherent complexity, I'm not sure I could do better :)

Maybe a few naming suggestions (as you see fit):
```
hasDoNotSetStatement => anyMethodHasOrDefaultsToDoNotSet
hasDoNotSetStatementSetExplicitly => anyMethodHasDoNotSetExplicitly
hasSetToNullStatementSetExplicitlyForAll => allMethodsHaveSetToNull (no need to say "explicitly" here, SET_TO_NULL is always explicit)

hasDaoLevelSetToNullStatement => daoHasSetToNull
hasDaoLevelDoNotSetStatement => daoHasDoNotSet
hasDaoLevelUnset => daoIsNotAnnotated
```

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (olim7t) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java]: I would move this as the main javadoc for the annotation. Also a few suggestions to reorganize the explanations:
```java
/**
 * Annotates a {@link Dao} interface to define a default {@link NullSavingStrategy}, that will apply
 * to all methods that don't explicitly declare one.
 *
 * <p>For example, given this interface:
 *
 * <pre>
 * &#64;Dao
 * &#64;DefaultNullSavingStrategy(SET_TO_NULL)
 * public interface ProductDao {
 *
 *   &#64;Insert
 *   void insert(Product product);
 *
 *   &#64;Update(nullSavingStrategy = DO_NOT_SET)
 *   void update(Product product);
 * }
 * </pre>
 *
 * <ul>
 *   <li>{@code insert(Product)} will use {@link NullSavingStrategy#SET_TO_NULL SET_TO_NULL}
 *       (inherited from the DAO's default).
 *   <li>{@code update(Product)} will use {@link NullSavingStrategy#DO_NOT_SET DO_NOT_SET}.
 * </ul>
 *
 * If the DAO interface isn't annotated with {@link DefaultNullSavingStrategy}, any method that does
 * not declare its own value defaults to {@link NullSavingStrategy#DO_NOT_SET DO_NOT_SET}.
 *
 * <p>Note that null saving strategies are only relevant for {@link Update}, {@link Insert}, {@link
 * Query} and {@link SetEntity} methods.
 */
```

**Review by olim7t — CHANGES_REQUESTED** [review]

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (tolbertam) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: What do you think about naming this simply `@NullSavingStrategy`?  Default seems kind of implicit, as to me it is intuitive to expect that if you were to also define `nullSavingStrategy` at the method-annotation level that it would override the interface level annotation.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:123]: There is a lot of boilerplate here between `queryHasDoNotSet`, `setEntityHasDoNotSet`, `insertHasDoNotSet` and `updateHasDoNotSet` that could probably be factored into a common method, for example you could do something like this:

```java
  private <A extends Annotation> boolean hasDoNotSet(Class<A> clazz,
                              Function<A, NullSavingStrategy> extractor,
                              ExecutableElement v,
                              boolean explicitSet) {
    A annotation = v.getAnnotation(clazz);
    if (annotation != null) {
      NullSavingStrategy strategy = extractor.apply(annotation);
      if (explicitSet) {
        return strategy == NullSavingStrategy.DO_NOT_SET
          && nullSavingStrategyExplicitlySet(v, clazz);
      } else {
        return strategy == NullSavingStrategy.DO_NOT_SET;
      }
    }
    return false;
  }

  private boolean queryHasDoNotSet(ExecutableElement v, boolean explicitSet) {
    return hasDoNotSet(Query.class, Query::nullSavingStrategy, v, explicitSet);
  }
```

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:157]: same here, these `HasSetToNullExplicitly` methods could be factored into a common method.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: If we were to remove `default NullSavingStrategy.DO_NOT_SET` from all annotations so they are null instead couldn't this be more effective as all you would have to do is null check?

File: manual/mapper/daos/defaultnullsavingstrategy/README.md
Comment (tolbertam) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: instead of `you can suppress defaults`, I would recommend `you can adjust the default behavior`

File: manual/mapper/daos/defaultnullsavingstrategy/README.md
Comment (tolbertam) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: These next two paragraphs don't add much as I think it should be apparent that if you only use one of the ways to configure null saving, the one that is used is used.

File: manual/mapper/daos/defaultnullsavingstrategy/README.md
Comment (tolbertam) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: If you agree and remove the two paragraphs below, you don't need this first line either.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java]: This code is repeated among the next four classes, could probably be refactored similarly to how I describe in `NullSavingStrategyValidation`

**Review by tolbertam — CHANGES_REQUESTED** [review]
Looks really great, lots of good test coverage.  I had some suggestions, but :+1: to the approach.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: Removing the default makes the attribute mandatory on the annotation.

Comment (tolbertam) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: ah, right.  That didn't occur to me at the time, in any case, could we just set default to null?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: Yes, I think that this is a good idea to have a "normal" java code where possible (instead of generated). Changed (https://github.com/datastax/java-driver/pull/1246/commits/37aada0af2153961c75d4fb7714b7b42ffc56ee4)

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:123]: I was also thinking about it - I had an impression that sometimes we are favoring code readability over DRY but If that is your impression I am +1 on this as well.
Implemented in https://github.com/datastax/java-driver/pull/1246/commits/0e99daa6a413387ae8478a01b5fd88c463edbc4f

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: If we will set it to `null` we will still need to interpret is somehow for a case where 
`@DefaultNullSavingStrategy` on the `@Dao` level is not a set AND the method level `nullSavingStrategy` property for example on `@Update` will not be set. 
In such case that null needs to become `DO_NOT_SET` or `SET_TO_NULL`. I think that it is better to have that default logic embedded within the annotation default:
`NullSavingStrategy nullSavingStrategy() default NullSavingStrategy.DO_NOT_SET;`
instead of adding another case to handle when the value is null.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java]: https://github.com/datastax/java-driver/pull/1246/commits/6d1e5e67a568b52db19fbc6a587288a9a593894e

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:98]: I like those names. Changed in https://github.com/datastax/java-driver/pull/1246/commits/7a40e48bbd042518390c5541b72ab67f4d7ae392

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (tomekl007) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: I agree with you that `@NullSavingStrategy` in this context should be enough.
The `@DefaultNullSavingStrategy` name was an initial idea of @olim7t - wdyt about changing that name to `@NullSavingStrategy`

File: manual/mapper/daos/defaultnullsavingstrategy/README.md
Comment (tomekl007) [reply manual/mapper/daos/defaultnullsavingstrategy/README.md]: The method level annotations have `NullSavingStrategy nullSavingStrategy() default NullSavingStrategy.DO_NOT_SET;` value. When the user is using it will have a default set.
Don't you think that some of our users may think that this unset default can take precedence?
I think that there is a small chance that someone may think like this so maybe it is better to have it explicitly documented?

**Review by tolbertam — APPROVED** [review]
:+1: my feedback has been addressed by recent changes.  Only question left is on the naming, but that is just a suggestion and i'm ok with either choice.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: In theory I would agree, but then the annotation would have the same name as the enum. I find this really unfortunate given that one references the other.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: Plus `null` is not a valid default for an annotation parameter.

**Review by olim7t — APPROVED** [review]

**Comment by olim7t** [comment]
Merged with `@DefaultNullSavingStrategy`: I felt there was a pretty strong case to not clash with the enum name.
I also took the liberty of rearranging the manual page.

**Pull request merged**
